code/datums/components/customizableatom.dm

```diff
@@ -0,0 +1,233 @@
+			filling.pixel_y = 2 * LAZYLEN(ingredients) - 1
+		if(CUSTOM_INGREDIENTS_STACKPLUSTOP)
+			filling.pixel_x = rand(-1,1)
+			filling.pixel_y = 2 *  LAZYLEN(ingredients) - 1
+			if (top_overlay) // delete old top if exists
+				P.cut_overlay(top_overlay)
+			top_overlay = mutable_appearance(P.icon, "[P.icon_state]_top")
+			top_overlay.pixel_y = 2 * LAZYLEN(ingredients) + 3
+			P.add_overlay(filling)
+			P.add_overlay(top_overlay)
+			return
+		if(CUSTOM_INGREDIENTS_FILL)
+			if (top_overlay)
+				filling.color = mix_color(filling.color)
+				P.cut_overlay(top_overlay)
+			top_overlay = filling
+		if(CUSTOM_INGREDIENTS_LINE)
+			filling.pixel_x = filling.pixel_y = rand(-8,3)
+	P.add_overlay(filling)
+
+
+///Takes the reagents from an ingredient.
+/datum/component/customizableatom/proc/handle_reagents(obj/item/I)
+	var/atom/P = parent
+	I.reagents.trans_to(P, I.reagents.total_volume)
+	return
+
+
+///Adds a new ingredient and its tastes and food types.
+/datum/component/customizableatom/proc/handle_ingredients(obj/item/I)
+	var/atom/P = parent
+	LAZYADD(ingredients, I)
+	P.name = "[custom_adjective()] [custom_type()] [initial(P.name)]"
+	SEND_SIGNAL(P, COMSIG_ATOM_CUSTOMIZED, I)
+
+
+///Gives an adjective to describe the size of the custom food.
+/datum/component/customizableatom/proc/custom_adjective()
+	switch(LAZYLEN(ingredients))
+		if (0 to 2)
+			return "small"
+		if (3 to 5)
+			return "standard"
+		if (6 to 8)
+			return "big"
+		if (8 to 11)
+			return "ridiculous"
+		if (12 to INFINITY)
+			return "monstrous"
+
+
+///Gives the type of custom food (based on what the first ingredient was).
+/datum/component/customizableatom/proc/custom_type()
+	var/custom_type = "empty"
+	if (LAZYLEN(ingredients))
+		var/obj/item/I = ingredients[1]
+		if (istype(I, /obj/item/food/meat))
+			var/obj/item/food/meat/M = I
+			if (M.subjectname)
+				custom_type = M.subjectname
+			else if (M.subjectjob)
+				custom_type = M.subjectjob
+		if (custom_type == "empty" && I.name)
+			custom_type = I.name
+	return custom_type
+
+
+///Returns the color of the input mixed with the top_overlay's color.
+/datum/component/customizableatom/proc/mix_color(color)
+	if(LAZYLEN(ingredients) == 1 || !top_overlay)
+		return color
+	else
+		var/list/rgbcolor = list(0,0,0,0)
+		var/customcolor = GetColors(color)
+		var/ingcolor =  GetColors(top_overlay.color)
+		rgbcolor[1] = (customcolor[1]+ingcolor[1])/2
+		rgbcolor[2] = (customcolor[2]+ingcolor[2])/2
+		rgbcolor[3] = (customcolor[3]+ingcolor[3])/2
+		rgbcolor[4] = (customcolor[4]+ingcolor[4])/2
+		return rgb(rgbcolor[1], rgbcolor[2], rgbcolor[3], rgbcolor[4])
+
+
+///Copies over the parent's ingredients to the processing result (such as slices when the parent is cut).
+/datum/component/customizableatom/proc/on_processed(datum/source, atom/original_atom, list/chosen_processing_option)
+	SIGNAL_HANDLER
+
+	var/datum/component/customizableatom/C = original_atom.GetComponent(/datum/component/customizableatom)
```

Comment:
You still need to find a better way of doing this that doesn't involve getcomponent